src/Microsoft.ML.HalLearners/VectorWhitening.cs

```diff
@@ -772,15 +671,143 @@ private static float DotProduct(float[] a, int aOffset, ReadOnlySpan<float> b, R
+                int maxRows = Defaults.MaxRows, int pcaNum = Defaults.PcaNum)
+            {
+                Name = name;
+                Contracts.CheckValue(Name, nameof(Name));
+                InputColumnName = inputColumnName ?? name;
+                Contracts.CheckValue(InputColumnName, nameof(InputColumnName));
+                Kind = kind;
+                Contracts.CheckUserArg(Kind == WhiteningKind.Pca || Kind == WhiteningKind.Zca, nameof(Kind));
+                Epsilon = eps;
+                Contracts.CheckUserArg(0 <= Epsilon && Epsilon < float.PositiveInfinity, nameof(Epsilon));
+                MaxRow = maxRows;
+                Contracts.CheckUserArg(MaxRow > 0, nameof(MaxRow));
+                SaveInv = Defaults.SaveInverse;
+                PcaNum = pcaNum; // REVIEW: make it work with pcaNum == 1.
+                Contracts.CheckUserArg(PcaNum >= 0, nameof(PcaNum));
+            }
+
+            internal ColumnInfo(VectorWhiteningTransformer.Column item, VectorWhiteningTransformer.Options options)
+            {
+                Name = item.Name;
+                Contracts.CheckValue(Name, nameof(Name));
+                InputColumnName = item.Source ?? item.Name;
+                Contracts.CheckValue(InputColumnName, nameof(InputColumnName));
+                Kind = item.Kind ?? options.Kind;
+                Contracts.CheckUserArg(Kind == WhiteningKind.Pca || Kind == WhiteningKind.Zca, nameof(item.Kind));
+                Epsilon = item.Eps ?? options.Eps;
+                Contracts.CheckUserArg(0 <= Epsilon && Epsilon < float.PositiveInfinity, nameof(item.Eps));
+                MaxRow = item.MaxRows ?? options.MaxRows;
+                Contracts.CheckUserArg(MaxRow > 0, nameof(item.MaxRows));
+                SaveInv = item.SaveInverse ?? options.SaveInverse;
+                PcaNum = item.PcaNum ?? options.PcaNum;
+                Contracts.CheckUserArg(PcaNum >= 0, nameof(item.PcaNum));
+            }
+
+            internal ColumnInfo(ModelLoadContext ctx)
+            {
+                Contracts.AssertValue(ctx);
+
+                // *** Binary format ***
+                // int:   kind
+                // float: epsilon
+                // int:   maxrow
+                // byte:  saveInv
+                // int:   pcaNum
+                Kind = (WhiteningKind)ctx.Reader.ReadInt32();
+                Contracts.CheckDecode(Kind == WhiteningKind.Pca || Kind == WhiteningKind.Zca);
+                Epsilon = ctx.Reader.ReadFloat();
+                Contracts.CheckDecode(0 <= Epsilon && Epsilon < float.PositiveInfinity);
+                MaxRow = ctx.Reader.ReadInt32();
+                Contracts.CheckDecode(MaxRow > 0);
+                SaveInv = ctx.Reader.ReadBoolByte();
+                PcaNum = ctx.Reader.ReadInt32();
+                Contracts.CheckDecode(PcaNum >= 0);
+            }
+
+            internal void Save(ModelSaveContext ctx)
```

Comment:
Are there save/load tests for this?

Reply:
in PcaTests.cs

---
In reply to: [254011887](https://github.com/dotnet/machinelearning/pull/2366#discussion_r254011887) [](ancestors = 254011887)